syntax = "proto3";

package order;

// Import payment service proto to use its enums
import "payment_service.proto";

service OrderService {
  // Create a new order
  rpc CreateOrder(CreateOrderRequest) returns (OrderResponse);
  
  // Get order details by ID
  rpc GetOrder(GetOrderRequest) returns (OrderResponse);
  
  // Update order status
  rpc UpdateOrderStatus(UpdateOrderStatusRequest) returns (OrderResponse);
  
  // Cancel an order
  rpc CancelOrder(CancelOrderRequest) returns (OrderResponse);
  
  // Get orders by customer ID
  rpc GetCustomerOrders(GetCustomerOrdersRequest) returns (OrderList);
  
  // Get orders by restaurant ID
  rpc GetRestaurantOrders(GetRestaurantOrdersRequest) returns (OrderList);
  
  // Update payment status for an order
  rpc UpdatePaymentStatus(UpdatePaymentStatusRequest) returns (OrderResponse);
}

message CreateOrderRequest {
  string customer_id = 1;
  string restaurant_id = 2;
  repeated OrderItem items = 3;
  string delivery_address = 4;
  string special_instructions = 5;
}

message OrderItem {
  string item_id = 1;
  string name = 2;
  int32 quantity = 3;
  double price = 4;
  repeated string customizations = 5;
}

message GetOrderRequest {
  string order_id = 1;
}

message UpdateOrderStatusRequest {
  string order_id = 1;
  OrderStatus status = 2;
  string notes = 3;
}

message CancelOrderRequest {
  string order_id = 1;
  string reason = 2;
}

message GetCustomerOrdersRequest {
  string customer_id = 1;
  int32 limit = 2;
  int32 offset = 3;
}

message GetRestaurantOrdersRequest {
  string restaurant_id = 1;
  OrderStatus status = 2;
  int32 limit = 3;
  int32 offset = 4;
}

message UpdatePaymentStatusRequest {
  string order_id = 1;
  string transaction_id = 2;
  payment.PaymentStatus payment_status = 3;
}

message OrderResponse {
  string order_id = 1;
  string customer_id = 2;
  string restaurant_id = 3;
  repeated OrderItem items = 4;
  double subtotal = 5;
  double tax = 6;
  double delivery_fee = 7;
  double total = 8;
  string delivery_address = 9;
  OrderStatus status = 10;
  payment.PaymentStatus payment_status = 11;
  string transaction_id = 12;
  string created_at = 13;
  string updated_at = 14;
  string special_instructions = 15;
}

message OrderList {
  repeated OrderResponse orders = 1;
  int32 total_count = 2;
}

enum OrderStatus {
  ORDER_PENDING = 0;
  ORDER_CONFIRMED = 1;
  ORDER_PREPARING = 2;
  ORDER_READY_FOR_PICKUP = 3;
  ORDER_OUT_FOR_DELIVERY = 4;
  ORDER_DELIVERED = 5;
  ORDER_CANCELLED = 6;
}